src/routes/object/getObjectInfo.ts

```diff
@@ -0,0 +1,147 @@
+import { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify'
+import { FromSchema } from 'json-schema-to-ts'
+import { IncomingMessage, Server, ServerResponse } from 'http'
+import { AuthenticatedRangeRequest, Obj } from '../../types/types'
+import { isValidKey, transformPostgrestError } from '../../utils'
+import { getConfig } from '../../utils/config'
+import { normalizeContentType } from '../../utils'
+import { createResponse } from '../../utils/generic-routes'
+import { S3Backend } from '../../backend/s3'
+import { FileBackend } from '../../backend/file'
+import { GenericStorageBackend } from '../../backend/generic'
+
+const { region, globalS3Bucket, globalS3Endpoint, storageBackendType } = getConfig()
+let storageBackend: GenericStorageBackend
+
+if (storageBackendType === 'file') {
+  storageBackend = new FileBackend()
+} else {
+  storageBackend = new S3Backend(region, globalS3Endpoint)
+}
+
+const getObjectParamsSchema = {
+  type: 'object',
+  properties: {
+    bucketName: { type: 'string', examples: ['avatars'] },
+    '*': { type: 'string', examples: ['folder/cat.png'] },
+  },
+  required: ['bucketName', '*'],
+} as const
+
+interface getObjectRequestInterface extends AuthenticatedRangeRequest {
+  Params: FromSchema<typeof getObjectParamsSchema>
+}
+
+async function requestHandler(
+  request: FastifyRequest<getObjectRequestInterface, Server, IncomingMessage>,
+  response: FastifyReply<
+    Server,
+    IncomingMessage,
+    ServerResponse,
+    getObjectRequestInterface,
+    unknown
+  >,
+  publicRoute = false
+) {
+  const { bucketName } = request.params
+  const objectName = request.params['*']
+
+  if (!isValidKey(objectName) || !isValidKey(bucketName)) {
+    return response
+      .status(400)
+      .send(createResponse('The key contains invalid characters', '400', 'Invalid key'))
+  }
+
+  const postgrest = publicRoute ? request.superUserPostgrest : request.postgrest
+  const objectResponse = await postgrest
+    .from<Obj>('objects')
+    .select('id')
+    .match({
+      name: objectName,
+      bucket_id: bucketName,
+    })
+    .single()
+
+  if (objectResponse.error) {
+    const { status, error } = objectResponse
+    request.log.error({ error }, 'error object')
+    return response.status(400).send(transformPostgrestError(error, status))
+  }
+
+  // send the object from s3
+  const s3Key = `${request.tenantId}/${bucketName}/${objectName}`
+
+  try {
+    const data = await storageBackend.headObject(globalS3Bucket, s3Key)
+
+    response
+      .status(data.httpStatusCode ?? 200)
+      .header('Content-Type', normalizeContentType(data.mimetype))
+      .header('Cache-Control', data.cacheControl)
+      .header('Content-Length', data.size)
+      .header('ETag', data.eTag?.replace(new RegExp('"', 'g'), ''))
```

Comment:
why is this regex required? in [other places](https://github.com/supabase/storage-api/blob/165197e03f48758e92728ca497a95b57f69ad522/src/routes/object/getPublicObject.ts#L89) looks like we just pass it through

Reply:
for some reason, the Etag that is coming from S3 is quoted "".
It felt a bit weird, but happy to return it as is if that's accetable